Source/Fuse.Controls.Native/iOS/InputDispatch.uno

```diff
@@ -8,131 +8,206 @@ namespace Fuse.Controls.Native.iOS
+						break;
+					case EVENTTYPE_MOVED:
+						@{InputDispatch.OnTouchesMoved(Visual,ObjC.Object):Call(visual, touches)};
+						break;
+					case EVENTTYPE_RELEASED:
+						@{InputDispatch.OnTouchesEnded(Visual,ObjC.Object):Call(visual, touches)};
+						break;
+					case EVENTTYPE_CANCELLED:
+						@{InputDispatch.OnTouchesCancelled(Visual,ObjC.Object):Call(visual, touches)};
+						break;
+					default:
+						break;
+				}
+			}, owner);
+		@}
 
-		static void DeactivateAllTouches()
-		{
-			_activeTouches.Clear();
-		}
+		[Foreign(Language.ObjC)]
+		public static void RemoveInputHandler(ViewHandle viewHandle)
+		@{
+			UIView* view = (UIView*)@{Fuse.Controls.Native.ViewHandle:Of(viewHandle).HitTestHandle:Get()};
+			removeInputHandler(view);
+		@}
 
-		public static void HandleEvent(ObjC.Object viewHandle, Visual origin, ObjC.Object uiEvent)
+		struct RootInfo
 		{
-			if (uiEvent.IsUIEvent())
-				HandleEvent(viewHandle, origin, new UIEvent(uiEvent));
+			public readonly Visual RootVisual;
+			public readonly ObjC.Object RootView;
+
+			public RootInfo(Visual rootVisual, ObjC.Object rootView)
+			{
+				RootVisual = rootVisual;
+				RootView = rootView;
+			}
 		}
 
-		public static void HandleEvent(ObjC.Object viewHandle, Visual origin, UIEvent uiEvent)
+		static RootInfo GetRootInfo(Visual origin)
 		{
 			var rootVisual = FindRoot(origin);
-			var touches = uiEvent.GetTouchesForView(viewHandle);
-			var touchEnded = false;
-
 			ObjC.Object rootView = null;
 			var rootViewport = rootVisual as NativeRootViewport;
 			if (rootViewport != null)
 				rootView = rootViewport.RootView.NativeHandle;
+			return new RootInfo(rootVisual, rootView);
+		}
 
-			for (var i = 0; i < touches.Length; i++)
-			{
-				var touch = touches[i];
+		const int MaxPointerCount = 10;
 
-				ActivateTouch(touch);
-				var pointerIndex = GetPointerIndex(touch);
-				var data = MakePointerEventData(touch, rootView, pointerIndex);
+		static ObjC.Object[] _activePointers = new ObjC.Object[MaxPointerCount];
 
-				if (touch.Phase == UITouch.TouchPhase.Began)
-				{
-					RaisePressed(rootVisual, origin, data);
-				}
-				else if (touch.Phase == UITouch.TouchPhase.Moved)
-				{
-					RaiseMoved(rootVisual, origin, data);
-				}
-				/*
-				else if (touch.Phase == UITouch.TouchPhase.Staionary) { }
-				*/
-				else if (touch.Phase == UITouch.TouchPhase.Ended)
-				{
-					RaiseReleased(rootVisual, origin, data);
-					touchEnded = true;
-				}
-				else if (touch.Phase == UITouch.TouchPhase.Cancelled)
-				{
-					RaiseCancelled(origin, data);
-					touchEnded = true;
-				}
+		static int GetPointIndex(ObjC.Object uiTouch)
+		{
+			var firstUnused = -1;
+			for (var i = 0; i < MaxPointerCount; ++i)
+			{
+				var pointer = _activePointers[i];
+				if (CompareUITouch(pointer, null) && firstUnused < 0)
+					firstUnused = i;
+				else if (CompareUITouch(pointer, uiTouch))
+					return i;
 			}
 
-			if (touchEnded)
+			_activePointers[firstUnused] = uiTouch;
```

Comment:
Minor point, it's highly unlikely, but maybe we should throw an exception if `firstUnused` is still -1. I can't get stats, but it's not impossible to imagine an iPad supporting more than 10 for multiple hands.